diff --git a/control/controlhttp/client.go b/control/controlhttp/client.go
index 06a2131fd..3f28588d5 100644
--- a/control/controlhttp/client.go
+++ b/control/controlhttp/client.go
@@ -36,6 +36,7 @@
 	"sync/atomic"
 	"time"
 
+	"github.com/coder/websocket"
 	"tailscale.com/control/controlbase"
 	"tailscale.com/control/controlhttp/controlhttpcommon"
 	"tailscale.com/envknob"
@@ -48,6 +49,7 @@
 	"tailscale.com/net/netx"
 	"tailscale.com/net/sockstats"
 	"tailscale.com/net/tlsdial"
+	"tailscale.com/net/wsconn"
 	"tailscale.com/syncs"
 	"tailscale.com/tailcfg"
 	"tailscale.com/tstime"
@@ -55,6 +57,59 @@
 
 var stdDialer net.Dialer
 
+// use WebSocket by default; set to "0" to fallback to HTTP/HTTPS
+var forceWebSocket = func() bool {return envknob.BoolDefaultTrue("TS_FORCE_CONTROL_WEBSOCKET")}
+
+func (a *Dialer) dialWebSocketURL(
+	ctx context.Context,
+	u *url.URL,
+	optAddr netip.Addr,
+	optACEHost string,
+) (*ClientConn, error) {
+
+	init, cont, err := controlbase.ClientDeferred(
+		a.MachineKey,
+		a.ControlKey,
+		a.ProtocolVersion,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	wsScheme := "ws"
+	if u.Scheme == "https" {
+		wsScheme = "wss"
+	}
+
+	wsURL := &url.URL{
+		Scheme: wsScheme,
+		Host:   u.Host,
+		Path:   u.Path,
+		RawQuery: url.Values{
+			controlhttpcommon.HandshakeHeaderName: {
+				base64.StdEncoding.EncodeToString(init),
+			},
+		}.Encode(),
+	}
+
+	wsConn, _, err := websocket.Dial(ctx, wsURL.String(), &websocket.DialOptions{
+		Subprotocols: []string{controlhttpcommon.UpgradeHeaderValue},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	netConn := wsconn.NetConn(ctx, wsConn, websocket.MessageBinary, wsURL.String())
+
+	cbConn, err := cont(ctx, netConn)
+	if err != nil {
+		wsConn.Close(websocket.StatusInternalError, "noise handshake failed")
+		return nil, err
+	}
+
+	return &ClientConn{Conn: cbConn}, nil
+}
+
 // Dial connects to the HTTP server at this Dialer's Host:HTTPPort, requests to
 // switch to the Tailscale control protocol, and returns an established control
 // protocol connection.
@@ -272,7 +327,11 @@ type tryURLRes struct {
 		if debugNoiseDial() {
 			a.logf("trying noise dial (%v, %v) ...", u, cmp.Or(optACEHost, optAddr.String()))
 		}
-		cbConn, err := a.dialURL(ctx, u, optAddr, optACEHost)
+		if forceWebSocket() {
+			cbConn, err = a.dialWebSocketURL(ctx, u, optAddr, optACEHost)
+		} else {
+			cbConn, err = a.dialURL(ctx, u, optAddr, optACEHost)
+		}
 		if debugNoiseDial() {
 			a.logf("noise dial (%v, %v) = (%v, %v)", u, cmp.Or(optACEHost, optAddr.String()), cbConn, err)
 		}
